walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,108 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    /**
+     * Given the duration for a meeting, a set of attendees are required
+     * to attend the meeting, and a set of events with their own respective 
+     * set of attendees, find and return the set of valid meeting times. 
+     * Valid meeting times are time ranges such that the meeting time is 
+     * greater than the given duration and all attendees are available 
+     * during that given time. 
+     */
+    public Collection<TimeRange> findMeetingTimes (
+        Collection<Event> events,    
+        Collection<String> attendees,
+        long duration) {
+
+        /* Sort events with latest end times coming first. */
+        List<Event> eventsByEndTime = new ArrayList<Event>(events);
+        List<Event> eventsByStartTime = new ArrayList<Event>(events);
+        Collections.sort(eventsByEndTime, Event.ORDER_BY_END_DESCENDING);
+        Collections.sort(eventsByStartTime, Event.ORDER_BY_START_ASCENDING);
+
+        Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+
+        /**
+         * To find valid meeting times, we use a temporary variable prevEndTime
+         * that is initially set to the end of the day. Since we sort events by 
+         * their end time in descending order, we effectively travel backwards.
```

Comment:
Something that is not clearly stated: why do we sort by the end time in descending order? Are there other choices? Is this the best one?

Reply:
This is interesting because in hindsight, both ways actually work with equal time complexity. I picked the second one because it made more logical sense to me, so I'm not quite sure how to phrase this as an explanation within a comment. (Adding this to Daily Sync because I have a few more questions)